### PR TITLE
Extra ACL checks (yay)

### PR DESCRIPTION
**Pull request by wilsonge**
Pull Request for Issue #11162 .
#### Summary of Changes

Adds bonus ACL checks for the categories on the fly
#### Testing Instructions

Before Patch:
- Create a new user with "manager" rights
- Remove the "Create" permission for managers in each component (com_content, com_contact, com_newsfeed and com_banners)
- => User can still create "categories on the fly" when editing an item

After Patch:
- Create a new user with "manager" rights
- Remove the "Create" permission for managers in com_content
- => User cannot  create "categories on the fly" when editing an item

### CODE REVIEW

File: administrator/components/com_newsfeeds/models/newsfeed.php [hunk 1]
```diff
@@ -503,9 +503,17 @@ protected function getReorderConditions($table)
 	 */
 	protected function preprocessForm(JForm $form, $data, $group = 'content')
 	{
+		// Check if article is associated
+		$canCreateCategories = JFactory::getUser()->authorise('core.create', 'com_newsfeeds');
```
Comment (izharaazmi) [inline administrator/components/com_newsfeeds/models/newsfeed.php]: Should/Does `core.create` on item also grants the permission to create categories for them?

Comment (andrepereiradasilva) [inline administrator/components/com_newsfeeds/models/newsfeed.php]: @izharaazmi go to Article -> Categories and click Options, what you see is the com_content permissions.
Same goes for all of the components with categories.

The question is if the user is allowed to create category items in the com_newsfeeds component.
So, correct. core.create - com_newsfeeds.

**Comment by infograf768** [comment]
Why:
`// Check if article is associated` everywhere?

File: administrator/components/com_banners/models/banner.php [hunk 2]
```diff
@@ -474,6 +474,30 @@ protected function prepareTable($table)
...
 	/**
+	 * Allows preprocessing of the JForm object.
+	 *
+	 * @param   JForm   $form   The form object
+	 * @param   array   $data   The data to be merged into the form object
+	 * @param   string  $group  The plugin group to be executed
+	 *
+	 * @return  void
+	 *
+	 * @since    3.0
```
Comment (andrepereiradasilva) [inline administrator/components/com_banners/models/banner.php]: 3.6.1 ?

Comment (wilsonge) [inline administrator/components/com_banners/models/banner.php]: Fixed ty!

File: administrator/components/com_contact/models/contact.php [hunk 3]
```diff
@@ -512,9 +512,30 @@ protected function getReorderConditions($table)
 	 * @param   string  $group  Group name.
 	 *
 	 * @return  void
+	 *
+	 * @since   3.0.3
 	 */
 	protected function preprocessForm(JForm $form, $data, $group = 'content')
 	{
+		// Determine correct permissions to check.
+		if ($this->getState('contact.id'))
```
Comment (wilsonge) [inline administrator/components/com_contact/models/contact.php]: This check is in all the other components - but missing from contacts so added this as well for consistency

**Comment by wilsonge** [comment]
Because I fail at copy/pasting code JM :) Fixed!

**Comment by izharaazmi** [comment]
I though **core.create - com_xyz** tells us whether the user is allowed to create item in **com_xyz**. But here I see the same permission is used to check whether if he is allowed to create **categories** for **com_xyz** items.

So how do we handle the practical case when the site admin has predefined set of categories and wants the authors to only create articles in those categories and NOT allow creating new categories. Am I missing on something?

**Comment by wilsonge** [comment]
Is that case handled at the moment in category creation?

**Comment by izharaazmi** [comment]
I'm not sure if this is possible in Joomla yet. I never did the site management so never had to go through that process. I just assumed it to be there because I believe this is so basic need, isn't it?

**Comment by andrepereiradasilva** [comment]
> I'm not sure if this is possible in Joomla yet. I never did the site management so never had to go through that process. I just assumed it to be there because I believe this is so basic need, isn't it?

since, for what i know, you're the first to ask doesn't seem a `basic need` :).

**Comment by izharaazmi** [comment]
Just to clarify my doubts.
Is it okay to allow authors to create an article category such as 'politics', 'cinema' etc. on my website dealing with 'medical science'? As I can see the authors do have 'core.create' on articles.
Am I really the first to ask this?

**Comment by mbabker** [comment]
If they have ACL permission to create items they can create items.  It's on the site admin to review that stuff, there isn't a way to programmatically state "my site is medical science so raise a red flag if someone starts creating art related content on the site".

> So how do we handle the practical case when the site admin has predefined set of categories and wants the authors to only create articles in those categories and NOT allow creating new categories.

Leave `core.create` at the default inherited state, create user groups with an ACL tree that doesn't explicitly allow (or deny) creating items at the component level, but specify they are allowed to create items within categories.  Unfortunately for your case IIRC that means they can create categories within that category as well as items, which is again the programatic correct behavior.  You'd need to extend the ACL system if you wanted to prevent the creation of specific item types even if the ACL allows the user to create items under a certain level.

**Comment by izharaazmi** [comment]
Thanks @mbabker. Now I can see what I was seeking for is not yet possible in Joomla.

> ... that means they can create categories within that category as well as items, which is again the programatic correct behaviour. 

However, it is still desirable to avoid allowing create everything in a component (or item ACL) based on a single rule `core.create` and delete everything in a component (or item ACL) with a single rule `core.delete`.

Similarly, in com_users we (super administrator) may not want to allow an administrator/manager to create user groups or  access level and still allow to create new users. Currently what I understand is we can't.

Wouldn't it be be something in the line of:
* `core.create` - `com_content` to allow content creation in articles component, 
* `core.category.create` - `com_content` for category creation in articles component.
and likewise in other relevant places. Hence giving more precise control on access over specific item type.

**Comment by mbabker** [comment]
> Wouldn't it be nice as something in the line of:

No.  `core.create` is a generalized permission level authorizing users to create items at a given level.  It does not know about or care about the different item types that can fall under that item, all it cares about is whether you've given the user permission to create stuff.  As a generalized solution, this is all Joomla really needs to care about.

I personally think bloating Joomla with permissions (i.e. `content.create.article` or `user.edit.group`, remember the `core.` "namespace" is reserved for what should be global actions) which may offer the most granular ACL control to a site is going to create too complicated of an interface for anyone to sanely manage their site's ACL.  Now assuming the ACL APIs are written correctly, programmatically you should be able to extend the base definitions and add more granular controls if you choose to do so.

At the end of the day, the current ACL is just fine, even with its known limitations, as a working starting point and generalized solution (because that's ultimately what the CMS application is, a compromise of a bunch of high level generalized solutions which aren't really optimized for anything) and the API should be strong enough to let folks extend it with additional custom rulesets if they really need a more fine tuned control set.

**Comment by izharaazmi** [comment]
Hmm... I understand the situation here. Agree to you.

Now as an expert advise would you please tell me whether I'm doing it right if I use following rules for my own extensions, or there are known better alternatives?

core.create = create everything.
component.create.abc = create item type abc
component.create.xyz = create item type xyz

and similar for delete, edit, state etc.

**Comment by mbabker** [comment]
That'd be the way to go about it, you'd just have to deal with the different possibilities that come out of that config set (i.e. if core.create is set to explicit deny following core's structure the component options should also be deny, or if you've allowed core.create but explicit denied one of the component options, things like that).

**Comment by izharaazmi** [comment]
Thanks again @mbabker. I appreciate the time and effort taken to explain the whole thing.

File: administrator/components/com_banners/models/banner.php [hunk 4]
```diff
@@ -474,6 +474,30 @@ protected function prepareTable($table)
...
+	 */
+	protected function preprocessForm(JForm $form, $data, $group = 'content')
+	{
+		// Check if article is associated
```
Comment (wojsmol) [inline administrator/components/com_banners/models/banner.php]: wrong comment

Comment (wilsonge) [inline administrator/components/com_banners/models/banner.php]: This got removed in the subsequent commit 5 hours ago :P

**Comment by bembelimen** [comment]
@wilsonge you have to add the ACL check to the saving process, too:

https://github.com/wilsonge/joomla-cms/blob/6d3d1293d88eed23528ef2193a5846215d6b8f38/administrator/components/com_categories/helpers/categories.php#L151-L152

and/or

https://github.com/wilsonge/joomla-cms/blob/6d3d1293d88eed23528ef2193a5846215d6b8f38/administrator/components/com_content/models/article.php#L508-L519 (and the other components)

**Comment by wilsonge** [comment]
@bembelimen Checks added to save method - sorry for the delay in adding them in and nice spot!

**Comment by brianteeman** [comment]
I have tested this item :white_check_mark: successfully on 0e7d0f8ea0b0dd179a568fed7a781e74acadf3b4<br /><br />Tested com_content, com_contact, com_newsfeed and com_banner - all good<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11244">issues.joomla.org/joomla-cms/11244</a>.</sub>

**Comment by MATsxm** [comment]
I have tested this item :white_check_mark: successfully on 0e7d0f8ea0b0dd179a568fed7a781e74acadf3b4<br /><br />Able to reproduce then works as described also tested for com_content, com_contact, com_newsfeed and com_banner<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11244">issues.joomla.org/joomla-cms/11244</a>.</sub>

**Comment by wilsonge** [comment]
Thanks! Merging my own PR so I can get an RC2 out

**Pull request merged**